Add Notifications for New Events

alexgraffeocohen opened the pull request:
This PR adds the ability for the Events Map to display notifications when new events are fetched that the Events Map doesn't already have. If there is only one new event, it will tailor the notification to the details of that event.

Also fixes a bug: we accidentally broke the "zoom into the event you just created" feature.

Some limitations of this feature:

* ~If the EventsMap is unmounted (i.e. the app is restarted in some way or somehow doesn't have enough memory to retain that screen) then we no longer know what events we previously had, so all fetched events are new again. We will ultimately rectify this with an `AsyncStorage` solution, but didn't want to go too far down that rabbit hole because the correct solution involves Redux. Could go there if needed, though. I think we should use [Redux Persist](https://github.com/rt2zz/redux-persist), which will allow us to store parts of our Redux state in `AsyncStorage`. This would be advantageous for other reasons, such as better offline support for this app so that users, even if they lose internet connection after the app is restarted, could still see the alerts they had already fetched.~
* We display a notification if the user themselves added a new event. To prevent that unneeded notification, we need to check for the currently signed in user and then see if the new event was created by them. For that matter, we could filter our all new events created by them, since they presumably know about them. This would be a pretty easy feature to add to this PR right now.
* ~Tapping the notifications does nothing. Navigating you to the map (if there are multiple new events) or the event show screen (if there is one new event) is a bit involved, but not impossible. Just didn't have time right now, and wanted a PR out the door.~

## ToDo

- [x] Store redux state in `AsyncStorage`
- [x] Navigate to the correct screens when tapping notifications
- [x] Present a modal explaining why we want notification permission before asking for it
- [ ] Optional: don't display notifications for events created by the current user
- [ ] Optional: don't display the snackbar if there are new events (we don't want snackbar and notification at the same time)

Commit:
Add iOS push notification entitlement

Commit:
Reference GoogleService-Info.plist file in Xcode

This file contains all of the information required by the Firebase iOS
SDK to connect to our Firebase project. We don't want to actually
include it in our repository because we are an open source project and
expect folks who fork the project to supply their own plist file.
However, Xcode still needs to know about the file.

The expected way for this file to be created is by constructing it via a
shell script and supplying the correct Firebase credentials using
environment variables. This commit demonstrates how this can work within
the project's appcenter-post-clone.sh script.

Create GoogleService-Info.plist file in App Center

Commit:
(m) delete whitespace

Commit:
Install Firebase core

* Add react-native-firebase so that we can use Firebase features within
our app
* Run `react-native link react-native-firebase`
* Install Firebase Cocoapod and add ios initialization code for Firebase
* Android will not work until we add the necessary android
initialization code, but since we don't have an Android app right now,
we're not doing that in this commit

Commit:
Install Firebase/Messaging pod

Commit:
Add iOS entitlement for getting notifications in background

Commit:
Configure notifications for iOS

Commit:
Remove detox from eslint env

Having it there breaks eslint because it's not a recognized env setting,
and so eslint was silently failing when trying to fix files in the
background.

Commit:
(m) Run ESLint

Commit:
(m) EventsMap: remove unused eventsServices

Commit:
(m) extract variable

Commit:
EventsMap: display notification on receiving new events

* Requests notification permission from App.js with default iOS dialog
for now. In the future, we should follow best practice and request
permission closer to where we need it and give the user context for
why we're asking.
* New events are determined by comparing old props and new props in
EventsMap.js. This means that when the app is restarted, we won't
remember what events we've already fetched. Future commits will improve
this by stashing events we've fetched already in local storage.
* Displays different notification depending on if there is one new event
or multiple

Commit:
EventsMap: zoom into newly added event again

We accidentally broke the feature of the EventsMap where if a user had
just created an event, we will zoom into the newly added event to show
them it was added correctly. Now, it will do that again!

Commit:
Remove linting of imports

It's been too complex to get working, so removing for now and will try
to get import linting working again at a later date.

Commit:
Extract store configuration out of App.js

Commit:
Add and configure redux-persist

This allows us to store our redux state in AsyncStorage so that if the
app loses memory, we will still retain the app state. This is
particularly important for events (and later alerts) because

1. It is important for users to retain downloaded data as long as
possible.
2. We will be much less likely to falsely notify users of new events
(and later alerts) that they had already seen.

Commit:
Navigate to correct screens on notification tap

This utilizes a technique for navigating to routes without access to the
navigator prop by stashing the ref for our navigator and proxying
navigate calls to it.

See https://reactnavigation.org/docs/en/navigating-without-navigation-prop.html

Commit:
Display explanatory alert before requesting notification permission

arghgr commented on ios/Redadalertas/GoogleService-Info.plist:
```diff
@@ -0,0 +1,36 @@
+	<key>REVERSED_CLIENT_ID</key>
+	<string>com.googleusercontent.apps.802628832645-6gbn0jhg1buk34kimru2ifj3h89c62bd</string>
+	<key>API_KEY</key>
+	<string>AIzaSyC4lz9mPm4JkJNWG91B9GUG2_yDOJpzUII</string>
```
I still have a feeling we should keep files like this out of version control, even if it's easily extracted from the APK or whatever. One of the reasons is to prevent people from gaining access to our Firebase, accidentally or otherwise, and also so any developers who want to work on it will be motivated to talk to us to get this information and we can see who they are.

Right now it sounds like it'll be difficult to both do that and have the file available for the distribution build. But long-term we should think about how to handle that. @thejoelpatrol how would this work for the Android distro?

alexgraffeocohen replied:
I'm fine getting this out of the repo for open source contribution reasons! Like I posted in slack, I would do something like what this script does in order to populate the plist file at build time from ENV variables in App Center:

```
#!/usr/bin/env bash

# Updating Info.plist

PLIST=$BUILD_REPOSITORY_LOCALPATH/MyApp/iOS/Info.plist

/usr/libexec/PlistBuddy -c "Set :CFBundleShortVersionString 1.0.${APPCENTER_BUILD_ID}" $PLIST

# Print out file for reference
cat $PLIST

echo "Updated info.plist!"
```

arghgr replied:
Oh ok sweet! I forgot we already talked about this. Maybe you can add that to the second notification PR (and then change the Firebase creds haha)

alexgraffeocohen replied:
Yeah I guess I could reset whatever credentials I can, but I won't be able to reset all of them (unless we create a totally different Firebase project). I'm happy to do it in this PR so that we at least never merge in the file.

arghgr replied:
Ok, that sounds good for now. We'll see if @thejoelpatrol has any security concerns once you reset what credentials you can.

arghgr commented on src/screens/Events/EventsMap.js:
```diff
@@ -92,8 +92,7 @@ class EventsMap extends Component {
       payload.action && payload.action.params ? payload.action.params : null;
-    if (params && params.refresh === true)
-      await this.props.getEvents(params.event || null);
+    if (params && params.refresh === true) await this.populateMap(params.event);
```
Thanks for catching this!

arghgr commented:
Ok, let me know once you're done pulling the Firebase creds out and we should be good to merge!
Also, another thing we might want to keep thinking about is, since you've added redux-persist (thanks for figuring that out!), is there any data in the Redux store that we *don't* want persisted on a user's device? I'll bring this up in the Slack.

Commit:
Remove detox tests

This was a solid experiment, but for whatever reason the detox
dependency started failing builds in App Center, so it's time for the
experiment to be over. We plan on switching end to end testing
frameworks anyways, so it's not a big loss.

alexgraffeocohen commented:
@arghgr OK we should be all set now! I removed the plist file from the repository, cleaned up commit history so it's as if I never added it, and added the app center build script to create the plist file during the build process. 

Like I said in Slack, all the info in that file is public, so I'm not going to try to reset the credentials, particularly since it's a whole thing (I'd have to delete the Firebase project and re-create it). 

I also removed detox, the end to end testing framework I tried, from the repo. It started messing up our App Center builds for some reason, and I want to replace it anyway.

The interesting commits to double check would be c13a0af and 7336638.

arghgr commented:
Ok, I've tested this PR and local notifications while the app is open are working, so I'm going to merge. Also added a reminder to add the gitignored file to the dev docs.
